lxd/devlxd.go

```diff
@@ -118,21 +118,43 @@ var devlxdEventsGet = devLxdHandler{"/1.0/events", func(d *Daemon, c instance.In
 		typeStr = "config,device"
 	}
 
-	conn, err := shared.WebsocketUpgrader.Upgrade(w, r, nil)
+	if r.Header.Get("Upgrade") == "websocket" {
+		conn, err := shared.WebsocketUpgrader.Upgrade(w, r, nil)
+		if err != nil {
+			return &devLxdResponse{"internal server error", http.StatusInternalServerError, "raw"}
+		}
+		defer conn.Close() // This ensures the go routine below is ended when this function ends.
+
+		listener, err := d.devlxdEvents.AddListener(c.ID(), conn, strings.Split(typeStr, ","))
+		if err != nil {
+			return &devLxdResponse{"internal server error", http.StatusInternalServerError, "raw"}
+		}
+
+		logger.Debug("New container event listener", logger.Ctx{"instance": project.Instance(c.Project(), c.Name()), "listener_id": listener.ID})
+		listener.Wait(r.Context())
+
+		return &devLxdResponse{"websocket", http.StatusOK, "websocket"}
+	}
+
+	h, ok := w.(http.Hijacker)
+	if !ok {
+		return &devLxdResponse{"internal server error", http.StatusInternalServerError, "raw"}
+	}
+
+	conn, _, err := h.Hijack()
 	if err != nil {
 		return &devLxdResponse{"internal server error", http.StatusInternalServerError, "raw"}
 	}
-	defer conn.Close() // This ensures the go routine below is ended when this function ends.
 
-	listener, err := d.devlxdEvents.AddListener(c.ID(), conn, strings.Split(typeStr, ","))
+	listener, err := d.devlxdEvents.AddListenerStream(c.ID(), conn, strings.Split(typeStr, ","))
```

Comment:
I think a comment here saying something along the lines of "If the client has not requested a websocket connection then fallback to long polling event stream mode".